modules/secp256k1/module.cpp

```diff
@@ -0,0 +1,160 @@
+#define template c_template // avoid C++ keyword conflict just during includes
+
+extern "C"
+{
+#include <secp256k1.h>
+#include <secp256k1_ecdh.h>
+}
+
+#undef template
+
+#include <iomanip>
+#include <sstream>
+#include <assert.h>
+#include "module.h"
+
+#define SECP256K1_PUBKEY_COMPRESSED_LEN 33
+#define SECP256K1_SIGNATURE_COMPACT_LEN 64
+#define SECP256K1_SIGNATURE_DER_MAX_LEN 74
+#define SECP256K1_SHARED_SECRET_LEN 32
+
+secp256k1_context *secp256k1_ctx;
+
+void init(int *argc, char ***argv)
+{
+	if (!secp256k1_ctx)
+	{
+		secp256k1_ctx = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
+		assert(secp256k1_ctx);
+	}
+}
+
+std::string hex_encode(const unsigned char *data, size_t len)
+{
+	std::ostringstream oss;
+	oss << std::hex << std::setfill('0');
+	for (size_t i = 0; i < len; ++i)
+	{
+		oss << std::setw(2) << static_cast<int>(data[i]);
+	}
+	return oss.str();
+}
+
+int create_pubkey_from_privkey(const uint8_t *privkey, secp256k1_pubkey &pubkey)
+{
+	int ret = 0;
+
+	ret = secp256k1_ec_seckey_verify(secp256k1_ctx, privkey);
+	ret = ret && secp256k1_ec_pubkey_create(secp256k1_ctx, &pubkey, privkey);
+
+	return ret;
+}
+
+int create_signature_from_hash(const uint8_t *privkey, const uint8_t *hash, secp256k1_ecdsa_signature &signature)
+{
+	int ret = 0;
+
+	ret = secp256k1_ec_seckey_verify(secp256k1_ctx, privkey);
+	ret = ret && secp256k1_ecdsa_sign(secp256k1_ctx, &signature, hash, privkey, nullptr, nullptr);
+	ret = ret && secp256k1_ecdsa_signature_normalize(secp256k1_ctx, &signature, &signature);
+
+	return ret;
+}
+
+std::optional<std::string> secp256k1_private_to_public(std::span<const uint8_t> buffer)
+{
+	int ret = 0;
+	secp256k1_pubkey pubkey;
+	size_t pubkey_len = SECP256K1_PUBKEY_COMPRESSED_LEN;
+	std::vector<uint8_t> pubkey_compressed(pubkey_len);
+
+	ret = create_pubkey_from_privkey(buffer.data(), pubkey);
+	ret = ret && secp256k1_ec_pubkey_serialize(secp256k1_ctx, pubkey_compressed.data(), &pubkey_len, &pubkey, SECP256K1_EC_COMPRESSED);
+	if (!ret)
+	{
+		return std::nullopt;
+	}
+
+	return hex_encode(pubkey_compressed.data(), pubkey_len);
+}
+
+std::optional<std::string> secp256k1_sign_compact(std::span<const uint8_t> buffer, std::span<const uint8_t> hash)
+{
+	int ret = 0;
+	secp256k1_ecdsa_signature signature;
+	std::vector<uint8_t> signature_compact(SECP256K1_SIGNATURE_COMPACT_LEN);
+
+	ret = create_signature_from_hash(buffer.data(), hash.data(), signature);
+	ret = ret && secp256k1_ecdsa_signature_serialize_compact(secp256k1_ctx, signature_compact.data(), &signature);
+	if (!ret)
+	{
+		return std::nullopt;
```

Comment:
Why it's returning `std::nullopt` here? In case it was not able to create the signature, for example, we shouldn't ignore it, right?

Reply:
I don't see a similar behaviour for the Decred module.